### ISSUE TITLE
Understand and document behaviour of stencil functions

### ISSUE THREAD

**Issue by lukasm91** [MeteoSwiss-APN/dawn#416]
DoD: 
    Document behaviour, expected behaviour and propose if moved to Frontend.

Before we can have anything like stencil functions in the unstructured code, we should clean up stencil functions for structured codes.

Currently, we have
```
message FieldAccessExpr {
     string name = 1;
     repeated int32 offset = 2;
     repeated int32 argument_map = 3;
     repeated int32 argument_offset = 4;
     bool negate_offset = 5;
 }
```
We don't see a need for `argument_map` and `argument_offset` being `repeated`. 

(Argument_map holds the index of the field that is passed to the stencil function, and offset is how it is accessed)

**Comment by cosunae** [comment]
say you have
```
stencil_function fn {
   storage in;
   direction dir1, dir2;
   Do {
     return in(dir1+1, dir2-1);
   }
};
```
then argument_map will be `{1,2}`
and argument_offset will be `{1,-1}`

therefore they can not be an int only, isnt?
At the same time, this functionality applies only to lat-lon grids probably, since on irregular grids we only have k left as a possible direction, unless we want to enable this on extra dimensions. But I guess we can think about that in the future. I would start assuming we only need this for lat-lon

**Comment by lukasm91** [comment]
For unstructured we would need to have a concept of a direcetion, and it would only work for vertical, so pretty meaningless.

When @twicki  and I tried to figure out what they actually do, our conclusion was that it will always only use one of the entries. Maybe then it would be a bug that you cannot write your second example. But also if we have
```
stencil_function fn {
   storage in;
   direction dir1, dir2;
   Do {
     return in(dir1+1, dir2-1);
   }
};
```
Do we make sure that `dir1 != dir2`? Maybe we can check again what actually works, and define what is the expected behavior.

**Comment by lukasm91** [comment]
Note that this task is not yet stencil functions for irregular grids! That might be a separate task (define a stencil function concept for irregular grid; no implementation required yet, but talking to GT4Py).

This task is really about this offsets. If you say that your example should work, I propose that we define this task as follow:
1. Document as-is state of stencil function concept for cartesian grids
2. Document expected behaviour of stencil function concept for cartesian grids 
3. Define action points to get from as-is to expected behavior.

**Comment by Stagno** [comment]
This is a nice feature of the frontend (having a `direction` variable type).
But what do we gain by also keeping it in dawn's AST? 
Although we already have code correctly handling `argument_map` and `argument_offset` in the backend, I see two problems: this code needs to be maintained and the SIR interface is more complicated than necessary.
Why not just convert it into an offset (`offset` of `FieldAccessExpr`'s protobuf) when translating from gtclang to SIR and get rid of `argument_map` and `argument_offset`?

**Comment by cosunae** [comment]
1 and 2 I can already answer here: 
as I described above the repeated is there because we have actually have mutliple directions in the field access. I think this is also what would be expected. 

In the case you mention where dir1=dir2, what it does today is accumulate all the offsets. 
i.e. `f(dir1+1, dir2-2)` when both `dir1` and `dir2` are `i`, it will become `i-1`
This is also what I would expect. This is also what happens in a field access with normal offsets, it does not imposes that the dimensions are different since it can resolve them. I see that you would give it a different semantic, so we can discuss it. But it is probably not related to this issue?

I also agree with @Stagno, but that would require coordination with the Frontends, so we should take it in a separate track of actions #433 

This task was to prepare the infrastructure to enable the stencil functions for irregular grids. Assuming that the arguments_xxx might dissappear in the future, what do we do ? 
We could just leave them empty when there is a stencil function on irregular grids, i.e. not use them (as it happens with any other field access that has no direction). 
That means we wouldnt need to do anything here ?

**Comment by lukasm91** [comment]
> This task was to prepare the infrastructure to enable the stencil functions for irregular grids. Assuming that the arguments_xxx might dissappear in the future, what do we do ?

No, this task was originally intended to clean up the stencil functions, completely independent form unstructured grids. I think if we have some intended behavior like the one you explain, the goal of this task should be adapted to document this properly.

> In the case you mention where dir1=dir2, what it does today is accumulate all the offsets.
i.e. f(dir1+1, dir2-2) when both dir1 and dir2 are i, it will become i-1
This is also what I would expect.

For me it's hard to imagine any meaningful behavior in the case both dimensions are i.
```
stencil_function fn {
  storage in; 
  direction dir1;
  Do { return in(dir1 + 1, i + 1); }
};
```
If I call this with `fn(in, i)`, I would probably even assume that this is simply not working because of ambiguity (it could mean `i+2` or `i+1`.

Then, this does not work:
```
stencil_function fn {
  storage in; 
  direction dir1;
  offset off1;
  Do { return in(dir1 + 1, off1); }
};
```
This function cannot be called. It might be reasonable to call it with `fn(i, j+1)`. Consequence of this is, that we can only call a function with one offset, or several directions (I think). In case of several directions, we would use `repeated argument_map`, but we will never use the `repeated` part of `argument_offset`.

Anyway, I feel like we are lacking proper documentation of this feature, and it feels like the protobuf specification does not well represent the feature.

>  We could just leave them empty when there is a stencil function on irregular grids, i.e. not use them (as it happens with any other field access that has no direction).

I agree with you that these arguments might just be optional. We can have a task for stencil functions for irregular grids (and we must involve GT4Py people there because this feature is driven by the frontend), we can do this probably also independent from this task, but still, we need to understand this feature, otherwise GT4Py cannot implement it.

**Comment by cosunae** [comment]
For reference, expected behaviour is the following: 

Any field access can only accept 3 arguments. If they are offsets, they are i,j,k
If they are dimension/direction parameters of a function, they can be anything and they can be repeated. 
`fied[dir1+1, dir2+3]` will turn into `field[i+4] ` when both dir1 and dir2 are instantiated as i

The only limitation, raised by lukas above is that offsets and direction/dimension arguments can not be combined. But this is a limitation only of gtclang that does not provide the corresponding overload, and not a dawn/SIR limitation

https://github.com/MeteoSwiss-APN/dawn/blob/bdd21771e6f8e1b8d99749ba21839642e146c139/gtclang/src/gridtools/clang/storage_dsl.hpp#L33-L56
